File: kyrex_engine/daemon.go
package kyrex_engine

// daemon.go — background engine daemon support for the TUI.
//
// The classic transport runs the engine as a child process wired over stdio:
// closing the TUI (or the terminal, or a crash) breaks the pipes and kills
// any in-flight turn. Daemon mode decouples the two lifecycles: the engine
// runs detached (KYREX_DAEMON=1), hosts a localhost TCP socket, buffers
// everything it emits for replay, auto-resolves approval gates while no UI
// is attached, and keeps working until an idle watchdog ends it. Quitting
// the TUI now just detaches — the session survives.
//
// The control-file key must match the Python side (daemon_bridge.py) and the
// IDE side (kyrex-ide/src-tauri/src/daemon.rs) exactly: FNV-1a 64 of the
// normalized workspace path, 16 lowercase hex chars.

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const (
	daemonConnectTimeout = 500 * time.Millisecond
	daemonStartTimeout   = 15 * time.Second
	daemonPollInterval   = 150 * time.Millisecond
)

// DaemonInfo mirrors the control file (~/.kyrex/daemons/<key>.json) written
// by daemon_bridge.py: {pid, port, workspace, project, started}.
type DaemonInfo struct {
	PID       int     `json:"pid"`
	Port      int     `json:"port"`
	Workspace string  `json:"workspace"`
	Project   string  `json:"project,omitempty"`
	Started   float64 `json:"started,omitempty"`
}

// normalizeDaemonWorkspace must mirror normalize_workspace in daemon_bridge.py
// and normalize_workspace in daemon.rs: separators to "/", trailing slashes
// trimmed, empty becomes "/".
func normalizeDaemonWorkspace(workspace string) string {
	p := strings.ReplaceAll(workspace, "\\", "/")
	p = strings.TrimRight(p, "/")
	if p == "" {
		return "/"
	}
	return p
}

// daemonFNV1a64 must mirror _fnv1a64 in daemon_bridge.py and fnv1a64 in
// daemon.rs. The "foobar" reference vector is covered by tests.
func daemonFNV1a64(data []byte) uint64 {
	const offset64 = uint64(0xcbf29ce484222325)
	const prime64 = uint64(0x100000001b3)
	h := offset64
	for _, b := range data {
		h ^= uint64(b)
		h *= prime64
	}
	return h
}

// DaemonKey returns the 16-hex-char control-file key for a workspace.
func DaemonKey(workspace string) string {
	return fmt.Sprintf("%016x", daemonFNV1a64([]byte(normalizeDaemonWorkspace(workspace))))
}

// daemonDir is where live daemons publish their {pid, port}. KYREX_HOME
// matches daemon_bridge.py's override; otherwise the user's home.
func daemonDir() string {
	if root := os.Getenv("KYREX_HOME"); root != "" {
		return filepath.Join(root, ".kyrex", "daemons")
	}
	home, err := os.UserHomeDir()
	if err != nil {
		home = os.Getenv("HOME")
	}
	return filepath.Join(home, ".kyrex", "daemons")
}

func daemonControlPath(workspace string) string {
	return filepath.Join(daemonDir(), DaemonKey(workspace)+".json")
}

// readDaemonInfo returns the recorded {pid, port} for this workspace, or nil.
func readDaemonInfo(workspace string) *DaemonInfo {
	raw, err := os.ReadFile(daemonControlPath(workspace))
	if err != nil {
		return nil
	}
	var info DaemonInfo
	if json.Unmarshal(raw, &info) != nil || info.Port <= 0 {
		return nil
	}
	return &info
}

// removeStaleDaemonControl prunes a control file whose port no longer
// answers, so later runs don't trust it.
func removeStaleDaemonControl(workspace string) {
	_ = os.Remove(daemonControlPath(workspace))
}

// listDaemonInfos parses every control file in the daemon dir. Corrupt or
// incomplete entries are skipped.
func listDaemonInfos() []*DaemonInfo {
	entries, err := os.ReadDir(daemonDir())
	if err != nil {
		return nil
	}
	var infos []*DaemonInfo
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		raw, err := os.ReadFile(filepath.Join(daemonDir(), e.Name()))
		if err != nil {
			continue
		}
		var info DaemonInfo
		if json.Unmarshal(raw, &info) != nil || info.Port <= 0 {
			continue
		}
		infos = append(infos, &info)
	}
	return infos
}

func dialDaemon(info *DaemonInfo) (net.Conn, error) {
	return net.DialTimeout("tcp", fmt.Sprintf("127.0.0.1:%d", info.Port), daemonConnectTimeout)
}

// spawnDetachedDaemon starts the engine with KYREX_DAEMON=1 as a process
// that deliberately outlives the TUI: own process group on Unix,
// DETACHED_PROCESS on Windows, stdio routed to a log file instead of the
// (short-lived) app.
func spawnDetachedDaemon(engineBin, pythonPath string, scriptArgs []string, workspaceRoot, projectSource string) error {
	var cmd *exec.Cmd
	if engineBin != "" {
		cmd = exec.Command(engineBin)
	} else {
		cmd = exec.Command(pythonPath, scriptArgs...)
	}

	key := DaemonKey(workspaceRoot)
	cmd.Env = append(os.Environ(),
		"KYREX_SURFACE=terminal",
		"KYREX_DAEMON=1",
		"WORKSPACE_ROOT="+workspaceRoot,
		"PROJECT_SOURCE_ROOT="+projectSource,
	)
	cmd.Dir = workspaceRoot

	logPath := filepath.Join(os.TempDir(), "kyrex-daemon-"+key+".log")
	logFile, err := os.OpenFile(logPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		return fmt.Errorf("daemon log: %w", err)
	}
	cmd.Stdin = nil
	cmd.Stdout = logFile
	cmd.Stderr = logFile

	configureDetached(cmd)

	if err := cmd.Start(); err != nil {
		logFile.Close()
		return fmt.Errorf("spawn daemon: %w", err)
	}
	// Reap whenever it eventually exits so it doesn't linger as a zombie.
	go func() {
		_ = cmd.Wait()
		logFile.Close()
	}()
	return nil
}

// AttachOrSpawnDaemon returns a Server wired to a background engine daemon
// for workspaceRoot, plus the directory the daemon is actually working in.
//
// Resolution order:
//  1. A daemon for exactly this workspace (same clone, or the IDE's daemon
//     on the project root itself).
//  2. A live daemon recorded for projectSource under a different clone root
//     — rift clones are recreated per run, so the workspace key changes
//     underneath a still-running daemon. Attaching to it lets the reopened
//     TUI pick up the live session instead of forking a second engine.
//  3. Spawn a fresh detached daemon and wait for its control file.
//
// The caller must adopt the returned workspace root when it differs from its
// own clone, so merge/sweep operate on the tree the daemon is editing. An
// error means daemon mode is unavailable — fall back to the classic child
// transport (which dies with the app).
func AttachOrSpawnDaemon(engineBin, pythonPath string, scriptArgs []string, workspaceRoot, projectSource string) (*Server, string, error) {
	// 1. Exact-workspace match.
	if info := readDaemonInfo(workspaceRoot); info != nil {
		if conn, err := dialDaemon(info); err == nil {
			return newDaemonServer(conn), normalizeDaemonWorkspace(info.Workspace), nil
		}
		removeStaleDaemonControl(workspaceRoot)
	}

	// 2. Project-level match (reattach across recreated clone roots).
	if projectSource != "" {
		if conn, info := findLiveProjectDaemon(projectSource); conn != nil {
			return newDaemonServer(conn), normalizeDaemonWorkspace(info.Workspace), nil
		}
	}

	// 3. Fresh detached daemon.
	if err := spawnDetachedDaemon(engineBin, pythonPath, scriptArgs, workspaceRoot, projectSource); err != nil {
		return nil, "", err
	}
	deadline := time.Now().Add(daemonStartTimeout)
	for time.Now().Before(deadline) {
		if info := readDaemonInfo(workspaceRoot); info != nil {
			if conn, err := dialDaemon(info); err == nil {
				return newDaemonServer(conn), normalizeDaemonWorkspace(info.Workspace), nil
			}
		}
		time.Sleep(daemonPollInterval)
	}
	return nil, "", fmt.Errorf(
		"engine daemon did not come up within %s (log: %s)",
		daemonStartTimeout,
		filepath.Join(os.TempDir(), "kyrex-daemon-"+DaemonKey(workspaceRoot)+".log"),
	)
}

// findLiveProjectDaemon scans the control files for a live daemon belonging
// to projectSource: newer control files record the project explicitly; older
// ones are matched by the workspace path living in that project's rift
// storage. Only daemons whose port actually answers are returned.
func findLiveProjectDaemon(projectSource string) (net.Conn, *DaemonInfo) {
	normalized := normalizeDaemonWorkspace(projectSource)
	clonesDir := ""
	if base := filepath.Base(projectSource); base != "" && base != "." && base != "/" {
		clonesDir = normalizeDaemonWorkspace(
			filepath.Join(filepath.Dir(projectSource), ".rifts", base))
	}
	for _, info := range listDaemonInfos() {
		if info.Project != "" {
			if normalizeDaemonWorkspace(info.Project) != normalized {
				continue
			}
		} else if clonesDir == "" ||
			!strings.HasPrefix(normalizeDaemonWorkspace(info.Workspace), clonesDir+"/") {
			continue
		}
		conn, err := dialDaemon(info)
		if err != nil {
			continue
		}
		return conn, info
	}
	return nil, nil
}

// newDaemonServer wraps a TCP connection to a live daemon in the same
// Server transport the TUI already speaks (newline-delimited JSON).
func newDaemonServer(conn net.Conn) *Server {
	return &Server{
		Stdin:  conn,
		Stdout: *bufio.NewReader(conn),
		conn:   conn,
		daemon: true,
	}
}

File: kyrex-ide/src-tauri/src/daemon.rs
//! daemon.rs — background engine discovery, spawning, and attachment.
//!
//! The engine runs in daemon mode (`KYREX_DAEMON=1`) as a detached process
//! that hosts a localhost TCP socket. Its lifecycle is decoupled from the
//! IDE: closing the app does not kill it, and a restarted IDE reattaches to
//! the live engine (replaying everything it missed) instead of spawning a
//! second one.
//!
//! The control-file key below MUST mirror `kyrex_engine/daemon_bridge.py`
//! (FNV-1a 64 of the normalized workspace path, 16 lowercase hex chars).

use serde::Deserialize;
use std::path::PathBuf;
use std::time::Duration;

#[derive(Debug, Clone, Deserialize)]
pub struct DaemonInfo {
    pub pid: u32,
    pub port: u16,
}

/// Must mirror `normalize_workspace` + `_fnv1a64` in daemon_bridge.py.
fn normalize_workspace(workspace: &str) -> String {
    let p = workspace.replace('\\', "/");
    let trimmed = p.trim_end_matches('/');
    if trimmed.is_empty() { "/".to_string() } else { trimmed.to_string() }
}

fn fnv1a64(data: &[u8]) -> u64 {
    let mut h: u64 = 0xcbf29ce484222325;
    for &b in data {
        h ^= b as u64;
        h = h.wrapping_mul(0x100000001b3);
    }
    h
}

pub fn daemon_key(workspace: &str) -> String {
    format!("{:016x}", fnv1a64(normalize_workspace(workspace).as_bytes()))
}

fn home_dir() -> Option<PathBuf> {
    std::env::var_os("HOME")
        .or_else(|| std::env::var_os("USERPROFILE"))
        .map(PathBuf::from)
}

pub fn control_file_path(workspace: &str) -> Option<PathBuf> {
    home_dir().map(|home| {
        home.join(".kyrex")
            .join("daemons")
            .join(format!("{}.json", daemon_key(workspace)))
    })
}

pub fn read_daemon_info(workspace: &str) -> Option<DaemonInfo> {
    let path = control_file_path(workspace)?;
    let raw = std::fs::read_to_string(path).ok()?;
    serde_json::from_str(&raw).ok()
}

pub fn remove_stale_control_file(workspace: &str) {
    if let Some(path) = control_file_path(workspace) {
        let _ = std::fs::remove_file(path);
    }
}

/// Resolves the bundled engine sidecar binary. Tauri places external binaries
/// next to the app executable, either bare ("kyrex-engine") or with the
/// platform target-triple suffix ("kyrex-engine-x86_64-unknown-linux-gnu").
pub fn resolve_engine_binary() -> Result<PathBuf, String> {
    let exe = std::env::current_exe().map_err(|e| format!("failed to resolve app exe: {e}"))?;
    let dir = exe
        .parent()
        .ok_or("app executable has no parent directory")?
        .to_path_buf();

    let exact = ["kyrex-engine", "kyrex-engine.exe"];
    for name in exact {
        let candidate = dir.join(name);
        if candidate.is_file() {
            return Ok(candidate);
        }
    }

    if let Ok(entries) = std::fs::read_dir(&dir) {
        let mut candidates: Vec<PathBuf> = entries
            .filter_map(|e| e.ok())
            .map(|e| e.path())
            .filter(|p| {
                let name = p.file_name().and_then(|n| n.to_str()).unwrap_or("");
                (name.starts_with("kyrex-engine-") && !name.ends_with(".d"))
                    || name == "kyrex-engine"
            })
            .collect();
        candidates.sort();
        if let Some(first) = candidates.first() {
            return Ok(first.clone());
        }
    }

    Err(format!(
        "kyrex-engine sidecar binary not found next to {}",
        exe.display()
    ))
}

/// Spawns the engine in detached daemon mode. The process deliberately
/// outlives the IDE: own process group on Unix, DETACHED_PROCESS on Windows,
/// stdio routed to a log file instead of the (short-lived) app.
pub fn spawn_detached_daemon(workspace: &str) -> Result<(), String> {
    let engine_bin = resolve_engine_binary()?;

    let key = daemon_key(workspace);
    let log_path = std::env::temp_dir()
        .join(format!("kyrex-daemon-{}.log", key));

    let mut cmd = std::process::Command::new(&engine_bin);
    cmd.current_dir(workspace)
        .env("KYREX_SURFACE", "Kyrex IDE")
        .env("KYREX_VSCODE", "1")
        .env("KYREX_DAEMON", "1")
        .env("WORKSPACE_ROOT", workspace);

    // Stdout + stderr both land in one append-mode log file (the tee'd
    // engine output plus tracebacks). One open, cloned for both streams.
    let log = std::fs::OpenOptions::new()
        .append(true)
        .create(true)
        .open(&log_path)
        .map_err(|e| format!("failed to open daemon log {log_path:?}: {e}"))?;
    let log_clone = log
        .try_clone()
        .map_err(|e| format!("failed to clone daemon log handle: {e}"))?;
    cmd.stdin(std::process::Stdio::null())
        .stdout(std::process::Stdio::from(log))
        .stderr(std::process::Stdio::from(log_clone));

    #[cfg(unix)]
    {
        use std::os::unix::process::CommandExt;
        cmd.process_group(0);
    }
    #[cfg(windows)]
    {
        use std::os::windows::process::CommandExt;
        const DETACHED_PROCESS: u32 = 0x0000_0008;
        const CREATE_NEW_PROCESS_GROUP: u32 = 0x0000_0200;
        cmd.creation_flags(DETACHED_PROCESS | CREATE_NEW_PROCESS_GROUP);
    }

    cmd.spawn()
        .map_err(|e| format!("failed to spawn engine daemon: {e}"))?;
    Ok(())
}

/// Waits (without blocking the async runtime) for the daemon to publish its
/// control file, then returns it.
pub async fn wait_for_daemon_info(workspace: &str, timeout_ms: u64) -> Result<DaemonInfo, String> {
    let deadline = tokio::time::Instant::now() + std::time::Duration::from_millis(timeout_ms);
    loop {
        if let Some(info) = read_daemon_info(workspace) {
            return Ok(info);
        }
        if tokio::time::Instant::now() >= deadline {
            return Err("engine daemon did not publish its control file in time".into());
        }
        tokio::time::sleep(Duration::from_millis(150)).await;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn key_matches_python_reference() {
        // Cross-checked against daemon_bridge._fnv1a64 / daemon_key.
        assert_eq!(daemon_key("/a/b"), daemon_key("/a/b/"));
        assert_eq!(daemon_key("C:\\repo\\sub"), daemon_key("C:/repo/sub"));
        assert_ne!(daemon_key("/ws/a"), daemon_key("/ws/b"));
        // FNV-1a 64 of "foobar" is a published reference vector.
        assert_eq!(fnv1a64(b"foobar"), 0x85944171f73967e8);
    }
}
